### PR TITLE
[Movement] (1/3): Movement Wizard

### PR DESCRIPTION
**Pull request by maltewae**
# Movement Update: Configuration Wizard and tests
Wizard to find, connect and adjust stages. Images can be find below.

## New names and concepts:
- `Mover`: entry point for all stage movements.
- `Calibration`: Manages a Stage (`Stage`-Object) and stores all necessary attributes to perform safe movements in LabExT. Such info are e.g. cooridnate axes calibrations.
- `Orientation`: is an enum of 4 different values: `LEFT`, `RIGHT`, `TOP`, `BOTTOM`. These values have no mathematical meaning, but allow the rough classification of the stage in the system and allow accesses like `left_calibration` etc. An orientation is assigned to a calibration (and thus to a stage). Globally there can be only one calibration for each orientation.
- `DevicePosition`: Is an enum of 2 different values: `INPUT`, `OUTPUT`. This value is assigned to a `Calibration` (and thus to a Stage) and indicates whether the Stage should move to the Input or Output position of the Device.

### MoverNew
(later this will replace the current Mover)The Mover class is the entry point for all stage movements in LabExTs. The task of the Mover is the following: It manages all connected Stages in LabExT. That means it manages a Dict `calibrations`, which assigns a tuple `(Orientation, DeviceInput)` to a `calibration`.
Important: The method `add_stage_calibration` creates new calibrations with given stage, position and orientation. This method ensures that no stage, no orientation and no position are used twice.
Example: `(LEFT, INPUT) -> Stage#1,` `(RIGHT, OUTPUT) -> Stage#1` (not possible because the same stage was used),
`(LEFT, INPUT) -> Stage#1`,` (LEFT, OUTPUT) -> Stage#2` (not possible, because the same orientation was used)
`(LEFT, INPUT) -> Stage#1`, `(RIGHT, INPUT) -> Stage#2`, (not possible because the same position was used)

The Mover also manages global stage settings, like speed and acceleration. It makes sure that all stages always have the same setting.

These properties are thoroughly tested in `MoverNew_test`.

### Calibration
Manages a stage and all properties for safe moving in LabExT (later here come the methods to move relative and absolute in chip-cords). In this PR only the attributes like orientation and position are set.

### MovementWizard
Wizard with 3 steps for the configuration of stages. First the drivers are loaded, then the stages can be assigned to an orientation and position as written above, then the global speeds and accelerations are tested.

The wizard is thoroughly tested in `MovementWizard_test`. There are unit tests for the wizard steps, unit tests for the controller and a big integration test, which tests a complete setup (complete run through of the wizard).

Screenshot 2022-02-14 at 17 13 07
Screenshot 2022-02-14 at 17 13 16
Screenshot 2022-02-14 at 17 13 25
Screenshot 2022-02-14 at 17 13 32

### CODE REVIEW

File: LabExT/View/MovementWizard/MovementWizardController.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        # Setting speed and acceleration values. Can throw StageError and
+        # ValueError.
+        if self.mover.has_connected_stages:
+            self.mover.speed_xy = speed_xy
+            self.mover.speed_z = speed_z
+            self.mover.acceleration_xy = acceleration_xy
+
+        # Refresh Context Menu
+        if self.experiment_manager:
+            self.experiment_manager.main_window.refresh_context_menu()
```
Comment (marcoep) [inline LabExT/View/MovementWizard/MovementWizardController.py:58]: Now I merged the PR #42 too early: there is already a feature to update the availability of the main-window check-boxes depending on the state of LabExT submodules. Its here: 

https://github.com/LabExT/LabExT/blob/e3e5d2d4c52a204c044f47223152f1d6a22bed7f/LabExT/View/MainWindow/MainWindowModel.py#L152

Maybe, you can add another PR to unify the functionality of `submodule_status_updated` and `refresh_context_menu` into something that tells the main GUI to "reload status from all submodules and adapt your GUI widgets accordingly"?

**Review by marcoep — COMMENTED** [review]
Thanks! Lets try it out tomorrow in the lab?

File: LabExT/View/MovementWizard/MovementWizardController.py (hunk 1 above)
Comment (maltewae) [reply LabExT/View/MovementWizard/MovementWizardController.py:58]: Good point, I did not see this method. I can combine them in a second PR.

File: LabExT/View/MenuListener.py [hunk 2]
```diff
@@ -183,6 +185,14 @@ def client_quit(self):
         """
         sys.exit(0)
 
+    def client_configure_mover(self):
+        """
+        Open wizard to configure the Mover.
+        """
+        if self.movement_wizard is None or not try_to_lift_window(self.movement_wizard.view):
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/View/MenuListener.py]: I updated the `MenuListener` such that `try_to_lift_window` will do a window check for every window.

File: LabExT/Movement/Calibration.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    LEFT = auto()
+    RIGHT = auto()
+    TOP = auto()
+    BOTTOM = auto()
+
+    def __str__(self) -> str:
+        return self.name.capitalize()
+
+
+class DevicePosition(Enum):
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Movement/Calibration.py]: I agree! I like `DevicePort`, I changed the naming.

File: LabExT/Movement/MoverNew.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+    #
+    #   Helpers
+    #
+
+    def _get_calibration(self, position=None, orientation=None, default=None):
+        """
+        Get safely calibration by position and orientation.
+        """
+        orientation = orientation or self._position_by_orientation.inverse.get(
```
[the remark this answers is not in the document]
Comment (maltewae) [reply LabExT/Movement/MoverNew.py]: It kind of hack:
```python
orientation = orientation or self._position_by_orientation.inverse.get(...)
```
will do the same as:
```python
orientation = orientation if orientation else self._position_by_orientation.inverse.get(...)
```

File: requirements.txt [hunk 5]
```diff
@@ -1,36 +1,29 @@
 #
-# This file is autogenerated by pip-compile with python 3.7
+# This file is autogenerated by pip-compile with python 3.8
 # To update, run:
 #
-#    pip-compile --output-file=requirements.txt '.\requirements.in'
```
[the remark this answers is not in the document]
Comment (maltewae) [reply requirements.txt]: I recompiled it for python 3.7

**Comment by maltewae** [comment]
@marcoep Should be use `dev` as branch or should be create an additional branch for all the movement changes?

**Comment by marcoep** [comment]
> ### **[maltewae](/maltewae) ** commented [6 hours ago](#issuecomment-1041311793)

Absolutely, yes. I see you have created some other branches here, are they to be used for this? Or should be create a new branch based on the current dev?

**Review by marcoep — APPROVED** [review]
Approving under the condition that we move it to another branch than dev.

**Comment by maltewae** [comment]
Changed the branch to `LabExT:movement/main`. The other branches are not up to date, I'll delete them.

**Pull request merged**